### **Group By**

---

In the prevouse section, we have learnt about joins and we are going to use it very  often

> **Joins are very important for visualization of data**

But what if we wanted to run anlytical data?

In the previouse sections, we have learnt about average, sum, count and etc for performing analytical operations

But what if we wanted to group our data into logical groups, let's say by department or by salary category or by gender 

There are just different ways of grouping data

---

> **Group by is a keyword that is used to summarize or aggregate data by groups.** 

Grouping data into logical buckets allows us to operate against those backets individually.

Why would we group data?

We group data to get in-depth information by group for instance the salary by department or the amount of employees by department or total amount of salary paid per department.

What if we wanted to know how many employees worked in each department?

Group by splits data into groups or chunks so we can apply functions against the group rahter that the entire table.


In [1]:
%load_ext sql

In [2]:
%sql postgresql://postgres:Kithusi254@localhost:5432/Employees

'Connected: postgres@Employees'

In [3]:
%%sql

SELECT table_name,
        string_agg(column_name, ' , ' ORDER BY ordinal_position)
FROM information_schema.columns
WHERE table_schema = 'public'
GROUP BY table_name
ORDER BY table_name

 * postgresql://postgres:***@localhost:5432/Employees
10 rows affected.


table_name,string_agg
cartesianA,id
cartesianB,id
departments,"dept_no , dept_name"
dept_emp,"emp_no , dept_no , from_date , to_date"
dept_manager,"dept_no , emp_no , from_date , to_date"
employees,"emp_no , birth_date , first_name , last_name , gender , hire_date"
salaries,"emp_no , salary , from_date , to_date"
staff_hierarchy,"staff_id , staff_name , role , manager_id"
timezones,"ts , tz"
titles,"emp_no , title , from_date , to_date"


In [4]:
%%sql

--What if we wanted to know how many employees worked in each department?

SELECT a.dept_name, 
COUNT(b.emp_no)
AS "no of emp in each department"
FROM dept_emp AS b 
INNER JOIN departments AS a ON b.dept_no = a.dept_no
GROUP BY a.dept_name

LIMIT 20;

 * postgresql://postgres:***@localhost:5432/Employees
9 rows affected.


dept_name,no of emp in each department
Customer Service,23580
Development,85707
Finance,17346
Human Resources,17786
Marketing,20211
Production,73485
Quality Management,20117
Research,21126
Sales,52245


---

The power of group by is, when we create these groups, the purpose of it is to run functions against some other parts of the data 

When we use group by, we are basically saying:

Take my data, split it into groups based on some column(s), then let me do calculations on each group.

Imagine we have a table of sales:

| customer | amount |
| -------- | ------ |
| John     | 100    |
| Mary     | 200    |
| John     | 50     |
| Mary     | 300    |

Without group by if we run:

**SELECT SUM(amount) FROM sales;**

We get a total of everthing, 650.

With group by

**SELECT customer, SUM(amount)
FROM sales
GROUP BY customer;**

Now SQL does 2 things:

* Creates groups:

* * Group 1 - John - (100, 50)
* * Group 2 - Mary - (200, 300)

* Runs a function on each group

* * John - SUM = 150
* * Mary - SUM = 500

So, GROUP BY organizes rows into groups

Then functions like:
* SUM()
* COUNT()
* AVG()
* MAX()
* MIN()

are applied inside each group, not on the whole table.

Key insight:

* Without group by - functions work on the entire dataset
* With group by - functions work on each group separately

---

**ANYTHING IN SELECT THAT IS NOT INSIDE A FUNCTION MUST BE IN GROUP BY.**

---

Group by splits data into groups or chunks so we can apply functions against the group rather than the entire table.

We use GROUP BY almost exclusively with aggregate functions because when we get a group of data we often want to know something about that group and that something we want to know, is most likely a sum, mean, count, etc

When we GROUP BY we apply the function per group, not on the entire data set

---

GROUP BY is stricter than it looks: 

We stated above that every column not in the GROUP BY clause must apply a function

How does GROUP BY work?

It utilizes a **Split-Apply-Combine strategy**

![](screen.png)

We split by the groups, we apply the aggregate functions to the remaining columns-columns we specify we need to apply agg functions and then we combine the results to get the output.

**Split phase**

Devide into groups with values

**Apply phase**

Apply aggregate against ungrouped columns

**Combine phase**

Combines groups with a single value into a single value like we see in the screenshot

---

### **Order of operations**

Earlier in the course we looked at the order of operations which is the order in which SQL does operations

**FROM - WHERE - GROUP BY - SELECT - ORDER**

GROUP BY happens after WHERE/FROM 

So if we have a WHERE clouse, the GROUP BY has to come after the WHERE clouse.

If we don't have the WHERE clouse, the GROUP BY has to come after the FROM clouse

And the rest of the process follows

The order of operation is important to know because we need to know where to put our GROUP BY 

If we put it in the wrong place, you will get errors

---

---

### **Having Keyword**

What if we want to filter groups?

This keyword is used after GROUP BY.

Syntax:

**SELECT col1, 
COUNT(col2)
FROM \<table\>
WHERE col2 > X
GROUP BY col1
HAVING col1 == Y**

The order of operation:

> **FROM - WHERE - GROUP BY - HAVING - SELECT - ORDER**

With the HAVING clouse, we can aplly filters aginst a group

Having works like a WHERE clouse

The difference is that WHERE applies filters to individual rows and HAVING applies filters to a group as a whole but not only the groups as a whole, when we apply agg functions against the group, the HAVING clouse can filter against the output of that agg function.

Let's do it practically...

---

In [5]:
%%sql

-- Let's say we wanted to know how many female employees we have in each
-- department in departments that have employees above 25000
-- We simply want to know which department has over 25000 female employees

SELECT table_name,
        string_agg(column_name, ' , ' ORDER BY ordinal_position)

FROM information_schema.columns
WHERE table_schema = 'public'
GROUP BY table_name
ORDER BY table_name;

 * postgresql://postgres:***@localhost:5432/Employees
10 rows affected.


table_name,string_agg
cartesianA,id
cartesianB,id
departments,"dept_no , dept_name"
dept_emp,"emp_no , dept_no , from_date , to_date"
dept_manager,"dept_no , emp_no , from_date , to_date"
employees,"emp_no , birth_date , first_name , last_name , gender , hire_date"
salaries,"emp_no , salary , from_date , to_date"
staff_hierarchy,"staff_id , staff_name , role , manager_id"
timezones,"ts , tz"
titles,"emp_no , title , from_date , to_date"


In [6]:
%%sql

SELECT d.dept_no,
        de.dept_name,
        COUNT(d.emp_no)
        AS "# of employees"
FROM employees AS e
INNER JOIN dept_emp AS d ON e.emp_no = d.emp_no
INNER JOIN departments AS de ON d.dept_no = de.dept_no
WHERE e.gender = 'F'
GROUP BY de.dept_name, d.dept_no
--Grouping by both dept_name and dept_id has no effect as they are the same.
HAVING COUNT(e.emp_no) > 25000
ORDER BY d.dept_no;

 * postgresql://postgres:***@localhost:5432/Employees
2 rows affected.


dept_no,dept_name,# of employees
d004,Production,29549
d005,Development,34258


---

When we try to use an aggregate function in a WHERE clouse, for instance, **WHERE COUNT(e.emp_no) > 20000**, we get errors that agg functions are not alowed in the WHERE.

The having clouse exist because of this exact reason.

Its entire purpose is to filter against the groups and it is allowed to use agg functions because it wants to filter against the singular output of that group

So when we use HAVING we can use the exact agg function that we do in our SELECT and we can filter against the specifics of the agg function, the output we are trying to get.

So for instance in our above example we know how many departments have 25000 plus youths

---

The difference between using WHERE and HAVING is that when you use having, if you are going to apply filter against a regular fieled rather than a group, well the it has to appear in you GROUB BY clouse because having applies against GROUPS and WHERE applies against each individual record

Let's illustrate this..

In [7]:
%%sql

SELECT table_name,
        string_agg(column_name, ' , ' ORDER BY ordinal_position)
FROM information_schema.columns
WHERE table_schema='public'
GROUP BY table_name
ORDER BY table_name;

 * postgresql://postgres:***@localhost:5432/Employees
10 rows affected.


table_name,string_agg
cartesianA,id
cartesianB,id
departments,"dept_no , dept_name"
dept_emp,"emp_no , dept_no , from_date , to_date"
dept_manager,"dept_no , emp_no , from_date , to_date"
employees,"emp_no , birth_date , first_name , last_name , gender , hire_date"
salaries,"emp_no , salary , from_date , to_date"
staff_hierarchy,"staff_id , staff_name , role , manager_id"
timezones,"ts , tz"
titles,"emp_no , title , from_date , to_date"


In [8]:
%%sql
--Departments that have female employees above 25000

SELECT de.dept_no,
        de.dept_name,
        COUNT(e.emp_no)
FROM employees AS e 
INNER JOIN dept_emp AS d ON e.emp_no = d.emp_no
INNER JOIN departments AS de ON d.dept_no = de.dept_no
GROUP BY de.dept_no, de.dept_name, e.gender
HAVING COUNT(e.emp_no) > 25000
-- For this filter to work, e.gender must be in the GROUP BY clouse 
AND e.gender = 'F';

 * postgresql://postgres:***@localhost:5432/Employees
2 rows affected.


dept_no,dept_name,count
d004,Production,29549
d005,Development,34258


---

So we can have a WHERE and HAVING in the same exact query

We see that we can do complex queries to get simple outputs.

For example how many employees are females and work in this department or males

---

---

### **Ordering Grouped data**

According to the order of operation, ORDER BY should come after GROUP BY

We can sort by eg, dept_no or by COUNT(e.emp_no), refering to previous examples

There are many ways to utilize ORDER BY but we can ORDER BY the groups we've already created

We can use the ORDER BY to run against the analytical functions that we have applied

**We can use our analytical functions in the  ORDER BY to ORDER BY output we get**

Let's do it practically....

---

In [9]:
%%sql

-- Let's look for the number of employees in each department and order by the count

SELECT table_name,
        string_agg(column_name, ' , ' ORDER BY ordinal_position)
FROM information_schema.columns
WHERE table_schema = 'public'
GROUP BY table_name
ORDER BY table_name

 * postgresql://postgres:***@localhost:5432/Employees
10 rows affected.


table_name,string_agg
cartesianA,id
cartesianB,id
departments,"dept_no , dept_name"
dept_emp,"emp_no , dept_no , from_date , to_date"
dept_manager,"dept_no , emp_no , from_date , to_date"
employees,"emp_no , birth_date , first_name , last_name , gender , hire_date"
salaries,"emp_no , salary , from_date , to_date"
staff_hierarchy,"staff_id , staff_name , role , manager_id"
timezones,"ts , tz"
titles,"emp_no , title , from_date , to_date"


In [10]:
%%sql

SELECT de.dept_no, 
        de.dept_name,
        COUNT(d.emp_no) 
FROM departments AS de
INNER JOIN dept_emp AS d ON de.dept_no = d.dept_no
GROUP BY de.dept_no, de.dept_name
ORDER BY COUNT(d.emp_no);



 * postgresql://postgres:***@localhost:5432/Employees
9 rows affected.


dept_no,dept_name,count
d002,Finance,17346
d003,Human Resources,17786
d006,Quality Management,20117
d001,Marketing,20211
d008,Research,21126
d009,Customer Service,23580
d007,Sales,52245
d004,Production,73485
d005,Development,85707


---

### **Group by mental model**

Get the most recent date that they got a salary bump

In GROUPING, we can ran our agg function to get some kind of answer but we can't ran our aggregate function to relate to columns of each data to each other in the same exact query

For instance in our case below, we got got the most recent date the emp had a salary increase but we couldn't find the salaries that correlate to the dates in the same query

Later we are going to learn how to do this

For now we need to understand this mental model of grouping data and drawing something out of it, an answer out of it

---

We also need to understand the shortcomings of GROUP BY.

Because we can grab an answer out of it, doesn't mean we can do everything in a singular query

THere are limitations to what we can do in a single query

But there will be techniques and ways we will learn down the way on how ro reuse that and chain it on other things 

Once we get an answer from a query, once we get data back, we can take that data and do other things 

The key thing to note is that after grouping data, the results we get we can reuse it later.

---

In [11]:
%%sql

-- Get the most recent date that they got a salary bump

SELECT emp_no,
        MAX(from_date)
FROM salaries
GROUP BY emp_no



LIMIT 20;

 * postgresql://postgres:***@localhost:5432/Employees
20 rows affected.


emp_no,max
10001,2002-06-22
10002,2001-08-02
10003,2001-12-01
10004,2001-11-27
10005,2001-09-09
10006,2001-08-02
10007,2002-02-07
10008,2000-03-10
10009,2002-02-14
10010,2001-11-23


---

### **Grouping sets**

In the previouse sessions we have seen how group by can trick us into a false sense that if we have grouping, the we can obviously calculate a result out of it.

So, grouping, although a very powerful tool, still has limitations, there are still some questions we cannot answer with group by. 

That was the purpose of the previouse section, to understand that even though grouping is a powerful tool, there are some questions we cannot answer with it.

What if we want to combine the results of multiple groupings?

How can we combine multiple answers from a group?

---

#### **UNION**

Union is used to combine results from multiple queries into one result 

It stacks results vertically (one on top of the other)

for eample:

SELECT name FROM employees
UNION
SELECT name FROM customers;

This gives us one list of names from both tables.

UNION is one of those powerful things things that when you want to run multiple queries and combine the results into one table, it can help you do that.

**Important rules**

1. Both queries must have:
    * same number of columns
    * compatible data types
2. UNION removes duplicates automatically.

#### **UNION ALL**

This is perfect if you want to retain duplicates

It does not remove duplicate records.

Sometimes you may come into scenarios where you need those duplicate records hence you can use UNION ALL

**Quick mental model**

* JOIN - combines columns (side by side)
* UNION - combines rows (top and bottom)

Let's do it practically...

---


In [12]:
%sql postgresql://postgres:Kithusi254@localhost:5432/Store

'Connected: postgres@Store'

In [13]:
%%sql

SELECT table_name,
        string_agg(column_name, ' , ' ORDER BY ordinal_position)
FROM information_schema.columns
WHERE table_schema='public'
GROUP BY table_name
ORDER BY table_name;

   postgresql://postgres:***@localhost:5432/Employees
 * postgresql://postgres:***@localhost:5432/Store
8 rows affected.


table_name,string_agg
categories,"category , categoryname"
cust_hist,"customerid , orderid , prod_id"
customers,"customerid , firstname , lastname , address1 , address2 , city , state , zip , country , region , email , phone , creditcardtype , creditcard , creditcardexpiration , username , password , age , income , gender"
inventory,"prod_id , quan_in_stock , sales"
orderlines,"orderlineid , orderid , prod_id , quantity , orderdate"
orders,"orderid , orderdate , customerid , netamount , tax , totalamount"
products,"prod_id , category , title , actor , price , special , common_prod_id"
reorder,"prod_id , date_low , quan_low , date_reordered , quan_reordered , date_expected"


In [14]:
%%sql
/*
Lets say for instance we wanted to select the grouped product 
ID together with there quantities and the total of there quantities,
We can use UNION to combine two queries to get the results we want
*/

SELECT 
        NULL AS "prod_id", -- This is important because UNION must have the same number of columns 
        --So we use NULL AS "prod_id" So that out query does not break.
        --We add a name for the column. If we didn't, the output would state no column.
        SUM(odl.quantity)
FROM orderlines AS odl

UNION

SELECT 
        odl.prod_id AS "prod_id",
        SUM(odl.quantity)
FROM orderlines AS odl
GROUP BY odl.prod_id 
ORDER BY prod_id DESC

LIMIT 50;


   postgresql://postgres:***@localhost:5432/Employees
 * postgresql://postgres:***@localhost:5432/Store
50 rows affected.


prod_id,sum
None,120719
10000,9
9999,13
9998,3
9997,16
9996,13
9995,13
9994,13
9993,8
9992,16


---

**Why NULL AS prod_id?**

Because both queries in a UNION must have:

* same number of columns
* matching structure

What's happening

second query returns:

**prod_id | SUM(quantity)**

So the first query must also return 2 columns

But there is no prod_id for the total

The first query is totaling everything

There is no specific products that is why we do **NULL AS prod_id**

Why NULL specifically?

* It clearly means "no value/not applicable"
* Better than putting fake ID like 0 (which could be missleading)

We could have done **'TOTAL' AS prod_id** but that only works if the column is text.

So in short, null is used to fill the missing column so UNION works, and to represent "no product"

---

**Deeper insight**

When we use UNION, SQL does this:

1. Runs both queries
2. Combines results into one dataset
3. Applies ORDER BY on the final product

So the only valid references are:

* Column names(prod_id)
* Column positions (ORDER BY 1) 

---

---

The whole point of these is to answe the question:

**What if we wanted to do all these in the same query?**

**What if we wanted to show the results of multiple queries in one query**

To achieve this, SQL gives us a powerful feature called **GROUPING SETS**

---

##### **GROUPING SETS**

GROUPING SETS allow us to combine the results of multiple queries kinda like the UNION would do.

It is a subclause of GROUP BY that allows us to define multiple groupings 

Let's do it practically...

---

In [15]:
%%sql

--Let's say we were to do the GROUPING SET of nothing and prod_id

SELECT prod_id,
        SUM(ol.quantity)
FROM orderlines AS ol
GROUP BY 
    GROUPING SETS(
        (),
        (prod_id)

    )
ORDER BY prod_id DESC
LIMIT 50;

   postgresql://postgres:***@localhost:5432/Employees
 * postgresql://postgres:***@localhost:5432/Store
50 rows affected.


prod_id,sum
None,120719
10000,9
9999,13
9998,3
9997,16
9996,13
9995,13
9994,13
9993,8
9992,16


---

If we run the above code, we can see that we are getting the same results.

It is just another way of rightint the UNION queries.

GROUPING SETS work the same way as a UNION would and its a powerful when you want to combine the outputs of multiple queries

We can go further and even do on different things like the order baseline id 

This would result to grouping three times when we add the order baseline id

This helps us to put it all in one query rather than having to write a union of multiple queries.

---

---

**Let's get orderid, prod_id and total quantity both from individual and grouped**

---

In [16]:
%%sql
SELECT table_name,
        string_agg(column_name, ' , ' ORDER BY ordinal_position)
FROM information_schema.columns
WHERE table_schema = 'public'
GROUP BY table_name
ORDER BY table_name

   postgresql://postgres:***@localhost:5432/Employees
 * postgresql://postgres:***@localhost:5432/Store
8 rows affected.


table_name,string_agg
categories,"category , categoryname"
cust_hist,"customerid , orderid , prod_id"
customers,"customerid , firstname , lastname , address1 , address2 , city , state , zip , country , region , email , phone , creditcardtype , creditcard , creditcardexpiration , username , password , age , income , gender"
inventory,"prod_id , quan_in_stock , sales"
orderlines,"orderlineid , orderid , prod_id , quantity , orderdate"
orders,"orderid , orderdate , customerid , netamount , tax , totalamount"
products,"prod_id , category , title , actor , price , special , common_prod_id"
reorder,"prod_id , date_low , quan_low , date_reordered , quan_reordered , date_expected"


In [17]:
%%sql

--Let's get orderid, prod_id and total quantity both from individual and grouped

SELECT 
    orderid, 
    prod_id, 
    SUM(quantity)
FROM orderlines
GROUP BY 
    GROUPING SETS(
        (),
        (prod_id),
        (orderid)
    )
ORDER BY prod_id DESC, orderid DESC  
LIMIT 50;

   postgresql://postgres:***@localhost:5432/Employees
 * postgresql://postgres:***@localhost:5432/Store
50 rows affected.


orderid,prod_id,sum
None,None,120719
12000,None,16
11999,None,15
11998,None,8
11997,None,21
11996,None,15
11995,None,10
11994,None,13
11993,None,13
11992,None,11


---

**What's happening?**

Let's break it down.

What our query is doing:

> **GROUP BY GROUPING SETS( (), (prod_id), (orderid))**

Were are telling postgreSQL, give me three different aggregations at ones

The three aggs are:

---

1. **()**

* orderid = NULL
* prod_id = NULL
* sum = NULL

This is the grand total(everything combined)

No grouping, so both columns become NULL

---

2. **(prod_id)**

* orderid = NULL
* prod_id = 6114
* sum = 14

This is the total quantity per product

Why is orderid NULL as when prod_id has a value we observed, because we are not grouping by orderid here.

---

3. **(orderid)**

* orderid = 12000
* prod_id = NULL
* sum = 16

This is the total quantity per order

Why is prod_id NULL while orderid has a value, because we are not grouping by prod_id

---

The reason as to why when prod_id has a value, orderid is NULL is because each row comes from a different grouping level, and any column bot used in that grouping becomes NULL.

**NOTE**

**In grouping sets, columns that are not part of the current grouping are filled with NULL**

---

Think of it like 3 seperate queries combined

Our query is basically doing:

**-- 1. Grand total**

SELECT NULL, NULL, SUM(quantity) FROM orderlines

UNION ALL

**-- 2. Per product**

SELECT NULL, prod_id, SUM(quantity)
FROM orderlines
GROUP BY prod_id

UNION ALL

**-- 3. Per order**

SELECT orderid, NULL, SUM(quantity)
FROM orderlines
GROUP BY orderid

---

So right now, NULL means two things:

* Not grouped by this column
* Actual missing data (if existed)

---

I realized that where prod id stops is where order id starts

* First: rows grouped bt prod_id → orderid = NULL
* Then: rows grouped by orderid → prod_id = NULL

The switch point is just where one grouping set ends and the other begins in the output like we see in the screenshot below

![](pic.png)

---

---

If we confirm if order 12000 has a total quantity of 16, we see we are right

Code below...

---

In [18]:
%%sql

SELECT SUM(quantity)
FROM orderlines
WHERE orderid = 12000;

   postgresql://postgres:***@localhost:5432/Employees
 * postgresql://postgres:***@localhost:5432/Store
1 rows affected.


sum
16


---
### **GROUPING(column)**

Ealier we said that NULL in GROUPING SETS can mean two things:

* Not grouped
* or actual missing data

GROUPING removes that confusion

It tells us whether that column was used in the grouping for that row

In our case below:

* **GROUPING(orderid) AS g_order**,
* **GROUPING(prod_id) AS g_prod**

It helps us know:

* Is this row per order?
* OR per product?
* Or grand total?

---

The output is interpreted as follows:

> **0 → column is used in grouping**

> **1 → column is not used in grouping**

---

In one line:

It's a lebal/flag to identify what level of agg each row belongs to.

Let's do it practically...

---

In [19]:
%%sql

SELECT table_name,
        string_agg(column_name, ' , ' ORDER BY ordinal_position)
FROM information_schema.columns
WHERE table_schema = 'public'
GROUP BY table_name
ORDER BY table_name;

   postgresql://postgres:***@localhost:5432/Employees
 * postgresql://postgres:***@localhost:5432/Store
8 rows affected.


table_name,string_agg
categories,"category , categoryname"
cust_hist,"customerid , orderid , prod_id"
customers,"customerid , firstname , lastname , address1 , address2 , city , state , zip , country , region , email , phone , creditcardtype , creditcard , creditcardexpiration , username , password , age , income , gender"
inventory,"prod_id , quan_in_stock , sales"
orderlines,"orderlineid , orderid , prod_id , quantity , orderdate"
orders,"orderid , orderdate , customerid , netamount , tax , totalamount"
products,"prod_id , category , title , actor , price , special , common_prod_id"
reorder,"prod_id , date_low , quan_low , date_reordered , quan_reordered , date_expected"


In [20]:
%%sql

/*
Let's get orderid, prod_id and total quantity both from individual and grouped
So we are getting the total quantity per order(grouped), total quatity per product(grouped) 
and total quatinties of individual quantities 
*/

SELECT 
        orderid, 
        prod_id, 
        SUM(quantity),
        GROUPING(orderid) AS g_order,
        GROUPING(prod_id) AS g_prod
FROM orderlines
GROUP BY
        GROUPING SETS(
                (),
                (orderid),
                (prod_id)
        )
ORDER BY prod_id DESC, orderid DESC
LIMIT 50;

   postgresql://postgres:***@localhost:5432/Employees
 * postgresql://postgres:***@localhost:5432/Store
50 rows affected.


orderid,prod_id,sum,g_order,g_prod
None,None,120719,1,1
12000,None,16,0,1
11999,None,15,0,1
11998,None,8,0,1
11997,None,21,0,1
11996,None,15,0,1
11995,None,10,0,1
11994,None,13,0,1
11993,None,13,0,1
11992,None,11,0,1


---

### **Rollup**



In [22]:
%%sql

SELECT  EXTRACT (YEAR FROM orderdate) AS "year",
        EXTRACT (MONTH FROM orderdate) AS "month",
        EXTRACT (DAY FROM orderdate) AS "day",
        SUM(ol.quantity),
        GROUPING(EXTRACT (YEAR FROM orderdate)) AS g_year,
        GROUPING(EXTRACT (MONTH FROM orderdate)) AS g_month,
        GROUPING(EXTRACT (DAY FROM orderdate)) AS g_date
FROM orderlines AS ol
GROUP By
    GROUPING SETS(
        --Group by year
        (EXTRACT (YEAR FROM orderdate)),
        (
        --Group by year and month
            EXTRACT (YEAR FROM orderdate),
            EXTRACT (MONTH FROM orderdate)
        ),
        (
        --Group by year, month and day
            EXTRACT (YEAR FROM orderdate),
            EXTRACT (MONTH FROM orderdate),
            EXTRACT (DAY FROM orderdate)
        ),
        (
        --Group by month and day
            EXTRACT (MONTH FROM orderdate),
            EXTRACT (DAY FROM orderdate)
        ),
        (EXTRACT (MONTH FROM orderdate)),
        (EXTRACT (DAY FROM orderdate)),
        () -- This gives us the total that we sold across all time
    )
ORDER BY 
    EXTRACT (YEAR FROM orderdate),
    EXTRACT (MONTH FROM orderdate),
    EXTRACT (DAY FROM orderdate)
LIMIT 100
;

   postgresql://postgres:***@localhost:5432/Employees
 * postgresql://postgres:***@localhost:5432/Store
100 rows affected.


year,month,day,sum,g_year,g_month,g_date
2004,1,1,329,0,0,0
2004,1,2,266,0,0,0
2004,1,3,315,0,0,0
2004,1,4,351,0,0,0
2004,1,5,420,0,0,0
2004,1,6,279,0,0,0
2004,1,7,300,0,0,0
2004,1,8,368,0,0,0
2004,1,9,347,0,0,0
2004,1,10,330,0,0,0


---

In the above query, we are trying to do the extraction of the year, the extraction of the month, and the extraction of the day and then we are doing the sum of the oderline quantities.

We are trying to figure out, how much were sold on the year, month and day available

But what if we wanted to know, how much did we sell on each day, and then how much did we sell on each month, and then how much did we sell on each year and we want it to be a pregressive set

Meaning we want to know each individual day by year and month,  and then we wanna know each individual month by year and we wanna know each individual year the total quantity that was sold.

So we want all that data

Grouping sets are perfect fro this because we are saying, heey, run multiple groupings and what we are trying to do here **is to group by year, year and month, and then year, month and day**

---

So, if we look at the output, we see, we are on the first month of 2004 and the first day of that year at our first record and we sold a total of 329 and so forth

As you go down, eventually we are going to hit the end of the month and we are going to see the total we sold for january of 2004, that is 10090 as we observe in our dataset

And we the continue to the second month and as we scroll down to the end of feb of 2004, we see we sold a total of 1075 for that month and so forward

---

**By Month and Day**

If you scroll down, you will reach to a place where we are now grouping by month and day alone as we see in the screenshot below.

![](Screenshot_2026-03-27_16-14-37.png)


---

**Yearly view**

If we go towards the bottom, what we are going to end up seeing is, instead of a monthly view or a daily view, we are going to see a yearly and daily view

![](pic2.png)

So on the first day of the month, no month, no year, we sold a total of 3967 as we see in the screenshot above, all the months combined from jan to dec.

On the second day of all the months combined we sold a total of 3754 and so on

---

**Total sold across all time**

As we move towards the end, we see the total we sold across all time.

![](pic4.png)


The total we sold across all time is 120, 719 as we see in our dataset.

---

So by grouping all this data this way, what we basicly have now is the ability to look at our data in a way that shows us multiples different type of groups and by knowing this, what we can do is kind of build those different groups together


---

---

#### **Rollup**

What we did up there is alot of work, we have to type out all the different groupings that we want but there is an easier way to do this and its called rollup

Rollup will do what we did up there.

Rollup is going to take our three inputs, and its going to make every possible combination between the three inputs we give it.

In our case above, the combination we got are not every available combination. If we chose to do every possible combination, it would need alot of code.

Let's do it...



In [23]:
%%sql

SELECT  EXTRACT (YEAR FROM orderdate) AS "year",
        EXTRACT (MONTH FROM orderdate) AS "month",
        EXTRACT (DAY FROM orderdate) AS "day",
        SUM(ol.quantity),
        GROUPING (EXTRACT (YEAR FROM orderdate)) AS g_year,
        GROUPING (EXTRACT (MONTH FROM orderdate)) AS g_month,
        GROUPING (EXTRACT (DAY FROM orderdate)) AS g_day
        
FROM orderlineS AS ol
GROUP BY
    GROUPING SETS(
        ROLLUP(
            EXTRACT (YEAR FROM orderdate),
            EXTRACT (MONTH FROM orderdate),
            EXTRACT (DAY FROM orderdate)
        )
    )
ORDER BY
    EXTRACT (YEAR FROM orderdate),
    EXTRACT (MONTH FROM orderdate),
    EXTRACT (DAY FROM orderdate)
--LIMIT 100
;

   postgresql://postgres:***@localhost:5432/Employees
 * postgresql://postgres:***@localhost:5432/Store
379 rows affected.


year,month,day,sum,g_year,g_month,g_day
2004,1,1,329,0,0,0
2004,1,2,266,0,0,0
2004,1,3,315,0,0,0
2004,1,4,351,0,0,0
2004,1,5,420,0,0,0
2004,1,6,279,0,0,0
2004,1,7,300,0,0,0
2004,1,8,368,0,0,0
2004,1,9,347,0,0,0
2004,1,10,330,0,0,0


---

So after we execute the code, we are going to see what we saw in our first code, Grouping by year month and day for every month through out the year and then at the end we have the total for 2004 by year, the total for 2004 by month and the total for 2004 by day and lastly the total for all time.

We have the following:

The total for:

* All time: 120719
* Every year: 120719
* By month: 9887
* Individual days: 389

So we created the same thing as before

The interesting thing about this is that it rolled up, that is why it is called rollup.

It dinamically creates this different groupings for us 

---

So, we are basically doing an extraction by year, month and day hence its rolling up.

Its doing all three, year, month and day, and then its doing two, month and day, and then its doing one, day, and then its doing what we knew to be NULL as we were starting with grouping set which is the grouping of all time.

It goes all the way up to nothing - NULL.

So its doing 4 groupings if i am not wrong:

1. YEAR, MONTH and DAY
2. MONTH and DAY
3. DAY
4. NULL

---

ROLLUP is usefull when we have grouping sets that need all the possible combinations between all the values 

It is an easier way to writing like that rather than writing all of the possible grouping sets, it it preaty error prone and you have to think of all the possible combinations.

---

How often we'll we use rollup?

Probably not super often depending on what you are trying to accomplish or what kind of questions been asked.

In our particular case is a good example, how much quantity did we sell across these three dates...

We will see it in practice depending on what industry you are working in

---

---

Exercise...

---

In [26]:
%sql postgresql://postgres:Kithusi254@localhost:5432/Employees

'Connected: postgres@Employees'

In [27]:
%%sql

SELECT table_name,
        string_agg(column_name, ' , ' ORDER BY ordinal_position)
FROM information_schema.columns
WHERE table_schema='public'
GROUP BY table_name
ORDER BY table_name;

 * postgresql://postgres:***@localhost:5432/Employees
   postgresql://postgres:***@localhost:5432/Store
10 rows affected.


table_name,string_agg
cartesianA,id
cartesianB,id
departments,"dept_no , dept_name"
dept_emp,"emp_no , dept_no , from_date , to_date"
dept_manager,"dept_no , emp_no , from_date , to_date"
employees,"emp_no , birth_date , first_name , last_name , gender , hire_date"
salaries,"emp_no , salary , from_date , to_date"
staff_hierarchy,"staff_id , staff_name , role , manager_id"
timezones,"ts , tz"
titles,"emp_no , title , from_date , to_date"


In [28]:
%%sql

/*
*  How many people were hired on any given hire date?
*  Database: Employees
*  Table: Employees
*/

SELECT  hire_date,
        COUNT(emp.emp_no)
FROM employees AS emp
GROUP BY hire_date
ORDER BY hire_date
LIMIT 20;

 * postgresql://postgres:***@localhost:5432/Employees
   postgresql://postgres:***@localhost:5432/Store
20 rows affected.


hire_date,count
1985-01-01,9
1985-01-14,1
1985-02-01,15
1985-02-02,110
1985-02-03,107
1985-02-04,106
1985-02-05,98
1985-02-06,98
1985-02-07,114
1985-02-08,118


In [29]:
%%sql

/*
*   Show me all the employees, hired after 1991 and count the amount of positions they've had
*  Database: Employees
*/

SELECT 
        emp.emp_no, 
        COUNT(ttl.title)
FROM employees AS emp
INNER JOIN titles AS ttl ON ttl.emp_no = emp.emp_no
--WHERE emp.hire_date > '1991/12/31'
GROUP BY emp.emp_no, hire_date
--For us to use having like this, filter on individual elements rather that groups, we need to 
--include it in the coulumn we are filtering by in the GROUP BY clouse.
HAVING emp.hire_date > '1991/12/31'
ORDER BY emp.emp_no
LIMIT 50;

 * postgresql://postgres:***@localhost:5432/Employees
   postgresql://postgres:***@localhost:5432/Store
50 rows affected.


emp_no,count
10008,1
10012,2
10016,1
10017,2
10019,1
10022,1
10024,1
10026,2
10030,2
10036,1


In [30]:
%%sql
-- Answer, easier way to do it

SELECT e.emp_no, count(t.title) as "amount of titles"
FROM employees as e
JOIN titles as t USING(emp_no)
WHERE EXTRACT (YEAR FROM e.hire_date) > 1991
GROUP BY e.emp_no
ORDER BY e.emp_no
LIMIT 50;

 * postgresql://postgres:***@localhost:5432/Employees
   postgresql://postgres:***@localhost:5432/Store
50 rows affected.


emp_no,amount of titles
10008,1
10012,2
10016,1
10017,2
10019,1
10022,1
10024,1
10026,2
10030,2
10036,1


In [31]:
%%sql

/*
*  Show me all the employees that work in the department development and the from and to date.
*  Database: Employees
*/

SELECT  emp.emp_no, 
        dpt.dept_no,
        dep.dept_name,
        dpt.from_date,
        dpt.to_date
FROM employees AS emp
INNER JOIN dept_emp AS dpt ON emp.emp_no = dpt.emp_no
INNER JOIN departments AS dep ON dpt.dept_no = dep.dept_no
WHERE dpt.dept_no = 'd005'
ORDER BY emp.emp_no
LIMIT 50;

 * postgresql://postgres:***@localhost:5432/Employees
   postgresql://postgres:***@localhost:5432/Store
50 rows affected.


emp_no,dept_no,dept_name,from_date,to_date
10001,d005,Development,1986-06-26,9999-01-01
10006,d005,Development,1990-08-05,9999-01-01
10008,d005,Development,1998-03-11,2000-07-31
10012,d005,Development,1992-12-18,9999-01-01
10014,d005,Development,1993-12-29,9999-01-01
10018,d005,Development,1987-04-03,1992-07-29
10021,d005,Development,1988-02-10,2002-07-15
10022,d005,Development,1999-09-03,9999-01-01
10023,d005,Development,1999-09-27,9999-01-01
10025,d005,Development,1987-08-17,1997-10-15


In [32]:
%%sql

SELECT e.emp_no, de.from_date, de.to_date
FROM employees as e
JOIN dept_emp AS de USING(emp_no)
WHERE de.dept_no = 'd005'
GROUP BY e.emp_no, de.from_date, de.to_date
ORDER BY e.emp_no, de.to_date

LIMIT 50;

 * postgresql://postgres:***@localhost:5432/Employees
   postgresql://postgres:***@localhost:5432/Store
50 rows affected.


emp_no,from_date,to_date
10001,1986-06-26,9999-01-01
10006,1990-08-05,9999-01-01
10008,1998-03-11,2000-07-31
10012,1992-12-18,9999-01-01
10014,1993-12-29,9999-01-01
10018,1987-04-03,1992-07-29
10021,1988-02-10,2002-07-15
10022,1999-09-03,9999-01-01
10023,1999-09-27,9999-01-01
10025,1987-08-17,1997-10-15


End......